[APM] Make sure URL hooks are in sync with history.location

dgieselaar opened the pull request:
Closes #36669.

Previously, the `useUrlParams` and `useLocation` hooks were possibly running out of sync with history.location, as they maintained their own state. By listening to history changes, they would re-render their context providers with the updated values. However, the context providers are wrapped by the `Router` component which re-renders the context providers when the location changes. This means that the initial render after a location change would have the updated values in `history.location`, but not in the values returned from the `useUrlParams` and `useLocation` hooks. These hooks would then (both) queue another render with the updated values being passed to the context.

This change lets the hooks piggyback on the re-renders from the `Router` component, and uses `history.location` to pass their derived values to the context. This ensures that the values returned from the hooks and `history.location` are always in sync. It should also remove a few unnecessary renders.

dgieselaar committed:
[APM] Make sure URL hooks are in sync with history.location

Previously, the `useUrlParams` and `useLocation` hooks were possibly running out of sync with history.location, as they maintained their own state. By listening to history changes, they would re-render their context providers with the updated values. However, the context providers are wrapped by the `Router` component which re-renders the context providers when the location changes. This means that the initial render after a location change would have the updated values in `history.location`, but not in the values returned from the `useUrlParams` and `useLocation` hooks. These hooks would then (both) queue another render with the updated values being passed to the context.

This change lets the hooks piggyback on the re-renders from the `Router` component, and uses `history.location` to pass their derived values to the context. This ensures that the values returned from the hooks and `history.location` are always in sync. It should also remove a few unnecessary renders.

Fix tests

- Passes history as a prop to URLParamsProvider, easier to test and consistent with usage of LocationProvider
- Do not test behaviour of history.replace & re-renders, only whether history.replace was called with the right Location values. This is because re-rendering is no longer a concern of URLParamsProvider, we leave that up to the Router component now.

Address review feedback:

- Use history/location from Router so we have one source of truth for routing info
- For UrlParamsContext, avoid use of history.replace. The behaviour w/ history.replace is incorrect. We now use useRef and trick React into re-rendering with an "empty" state update.
- Unit test for whether useUrlParams does not queue an unnecessary render.

Memoize context value of UrlParamsProvider

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13979/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14009/)

sqren commented on x-pack/plugins/apm/public/context/LocationContext.tsx:
```diff
@@ -15,15 +15,9 @@ const initialLocation = {} as Location;
-  }, []);
+  const location = {
+    ...history.location
+  };
```
Why create a new `location` object on every render? Why not just pass `history.location` through?

dgieselaar replied:
`history.location` is mutable. I figure it's easy to make a copy here and prevent any issues w/ some other consumer accidentally mutating the `history.location` object downstream.

sqren replied:
> history.location is mutable

Not the one coming from react router:
https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/history.md#history-is-mutable

The problem with this is that it will always cause consumers of LocationContext to update when _any_ parent of LocationContext re-renders (even if `location` stays the same)

sqren replied:
Oh I just saw `<LocationProvider history={history}>` is already like that - yeah, I don't think we should do that. Instead we should just get `location` which `Router` will pass as a prop to child components (and which will be immutable)

sqren commented on x-pack/plugins/apm/public/context/UrlParamsContext/index.tsx:
```diff
@@ -4,74 +4,40 @@
+        end: time.rangeTo
+      })
+    });
+  };
```
I think this will break the "refresh" button. If the user has a relative time range `rangeFrom` and `rangeTo` will not change and so the url will not change.

sqren replied:
Btw. `urlParams ` are derived from the url so should only ever read from it - not change it.

dgieselaar replied:
I tested that, and it seems to still work. It basically just refreshes the current URL, triggering a re-render via the `Router` component, and then the start/end values are re-calculated again.

sqren replied:
I think this could just be an optimization away from breaking. It might be because whenever react-router updates, all context providers are currently not using memo as suggested in https://github.com/elastic/kibana/pull/35939. 

Eg. if the relative date hasn't changed we don't re-calculate the absolute date:
https://github.com/elastic/kibana/blob/3efd42c63b67f6d8ba11a79abd6e087d457279ba/x-pack/plugins/apm/public/context/UrlParamsContext/helpers.ts#L21-L24

sqren commented on x-pack/plugins/apm/public/new-platform/plugin.tsx:
```diff
@@ -32,7 +32,7 @@ function App() {
 
   return (
-    <UrlParamsProvider>
+    <UrlParamsProvider history={history}>
```
Any reason not to get `location` via the `useLocation` hook?

dgieselaar replied:
It needs the whole `history` object, as it calls `history.replace()` (in `refreshTimeRange`).

dgieselaar commented:
As discussed with @sqren, we removed the use of `history.replace` in `UrlParamsContext` with some trickery w/ useRef and forced updates. Would like to iterate on this solution, but could not figure out a better way that:

- does not render incorrectly on the first location change
- does not queue an unnecessary render

@ogupte maybe you have an idea on how to improve?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14764/)

sqren commented on x-pack/plugins/apm/public/context/LocationContext.tsx:
```diff
@@ -4,28 +4,17 @@
-  return <LocationContext.Provider children={children} value={location} />;
-};
+const LocationProvider: React.ComponentClass<{}> = withRouter(
+  ({ location, children }) => {
```
Since we now have access to react-router we could also pass `matches` along. That would be very useful for `resolveUrlParams` so we can get rid of `getPathParams`

dgieselaar replied:
@sqren not a react-router expert, but it seemingly only works when the component that uses the `withRouter` HOC is below a `Route`. The `LocationProvider` is below a `Router`, but not a `Route`, so `match` returns an object with a bunch of default (useless) values.

sqren replied:
> The LocationProvider is below a Router, but not a Route, so match returns an object with a bunch of default (useless) values.

Ahh, that rings a bell. Well, nvm then.

sqren commented on x-pack/plugins/apm/public/context/UrlParamsContext/index.tsx:
```diff
@@ -40,58 +35,49 @@ function useUiFilters(urlParams: IUrlParams): UIFilters {
+    return (
+      <UrlParamsContext.Provider
+        children={children}
+        value={{ urlParams, refreshTimeRange, uiFilters }}
```
What do you think about wrapping this in `useMemo` so it doesn't compute a new object on every render?

dgieselaar replied:
Aye, will fix.

sorenlouv reviewed (approved):
LGTM.
Please test that both relative and absolute ranges work as expected, and that this doesn't cause duplicate xhr requests to be fired.

dgieselaar commented:
@sqren I think the time range scenarios are fully covered by tests. For firing XHR requests, I've clicked around for a bit and everything seems ok. Do you have any specific scenarios I should look at?

sqren commented:
> Do you have any specific scenarios I should look at?

No, sounds like everything is taken care of 👍

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15261/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15284/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15316/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15455/)

dgieselaar committed:
Move infra items back to the top again

Piggybacking on this change to fix an order issue in TransactionActionMenu that was introduced w/ 2e3fbd24a8fee57bb87eafd536fc3c6ed8c6d81b (woops).

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15459/)
